File: api/src/modules/power/power.service.ts
import prisma from '../../lib/prismaClient';
import { AppError } from '../../middleware/errorHandler';

class PowerService {
  async getCategories() {
    const categories = await prisma.powerCategory.findMany({
      where: { isActive: true },
      orderBy: { sortOrder: 'asc' },
    });

    return categories;
  }

  async saveSelections(userId: number, categoryIds: number[]) {
    return await prisma.$transaction(async (tx) => {
      // Delete existing selections
      await tx.userPowerSelection.deleteMany({
        where: { userId },
      });

      // Create new selections
      const selections = await tx.userPowerSelection.createMany({
        data: categoryIds.map((powerCategoryId, index) => ({
          userId,
          powerCategoryId,
          priority: index + 1,
        })),
      });

      return { message: 'Selections saved successfully', count: selections.count };
    });
  }

  async getSelections(userId: number) {
    const selections = await prisma.userPowerSelection.findMany({
      where: { userId },
      orderBy: { priority: 'asc' },
      include: {
        powerCategory: true,
      },
    });

    return selections;
  }
}

export default new PowerService();

File: mobile/lib/services/power_service.dart
import '../models/power_category.dart';
import 'api_client.dart';

class PowerService {
  final ApiClient _client;

  PowerService(this._client);

  Future<List<PowerCategory>> getCategories() async {
    final json = await _client.get('/power/categories');
    final data = json['data'] as List;
    return data.map((c) => PowerCategory.fromJson(c as Map<String, dynamic>)).toList();
  }

  Future<void> saveSelections(List<int> categoryIds) async {
    await _client.post('/power/selections', body: {
      'categoryIds': categoryIds,
    });
  }

  Future<List<UserPowerSelection>> getSelections() async {
    final json = await _client.get('/power/selections');
    final data = json['data'] as List;
    return data.map((s) => UserPowerSelection.fromJson(s as Map<String, dynamic>)).toList();
  }
}

